connman/src/service.c

```diff
@@ -510,10 +512,49 @@ static inline char *service_path(const char *ident)
 	return g_strconcat(CONNMAN_PATH, "/service/", ident, NULL);
 }
 
+
+static void count_available_service_type(struct connman_service *service,
+								bool add)
+{
+	GHashTable *table;
+	enum connman_service_type type;
+
+	if (!service)
+		return;
+
+	type = service->type;
+	table = service_type_hash[type];
+
+	DBG("%p/%s %s %c 1", service, service->name,
+					__connman_service_type2string(type),
+					add ? '+' : '-');
+
+	if (add) {
+		if (!is_available(service)) {
+			DBG("service %p/%s not available", service,
+							service->identifier);
+			return;
```

Comment:
If a service is not going to be counted "in", should it be counted "out"? In if(add && available) insert(); else remove(); sort of way?

Reply:
The way different plugins seem to work: very differently they can try to add them in places before and after they are available. It is somewhat of a mess and this is just a guarantee to operate (hopefully) properly in all cases by just ignoring the addition in cases where it is called prior to being available.

Kind of difficult to think of a unit test for this either as the plugins just can work with different needs...

Reply:
But this made me think if the `bool` is necessary after all. I tried this with `int` arrays first but with this hash table things are different..

Reply:
Well, yeah, now I remember. VPN services and others alike that have different availability requirements do not work with plain availability check as they are always available, so removal does not work when relying on the availability. This is complex..

Reply:
So, once a service has been counted as "added and available", it is ok for it to go back and forth between available and not-available without the count changing until service removal?

Reply:
That does not happen, the `not available` is the case when loading the service, when a service has its network removed the count is decreased (any service with a changing network condition) or removed, which includes also the VPNs that are removed (has provider).

Reply:
As it seemed that in some cases when connecting directly to a new network it can be in available state when it is first counted (or this is what I remember happening while testing) but in most of the cases it is not available when creating. Same function seems to get called twice in some cases, with same conditions except the network availability changes...

Reply:
Ok. I have no idea what the pig picture is here and it just looked a bit suspicious. ..